drivers/hid/hid-multitouch.c

```diff
@@ -1673,6 +1691,59 @@ static void mt_expired_timeout(struct timer_list *t)
 	clear_bit(MT_IO_FLAGS_RUNNING, &td->mt_io_flags);
 }
 
+static void update_keyboard_backlight(struct hid_device *hdev, bool enabled)
+{
+	struct usb_device *udev = hid_to_usb_dev(hdev);
+
+	unsigned char cmd_disable_backlight[] =
+		{0x24, 0xFF, 0x00, 0xFF, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
+	unsigned char cmd_enable_backlight[] =
+		{0x24, 0xFF, 0x00, 0xFF, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
+
+	unsigned char *cmd_buf;
+	unsigned char *dmabuf;
+	int ret;
+
+	/* Wake up the device in case it's already suspended */
+	if (pm_runtime_suspended(&udev->dev))
+		pm_runtime_resume(&udev->dev);
+
+	cmd_buf = enabled ? cmd_enable_backlight : cmd_disable_backlight;
+
+	dmabuf = kmemdup(cmd_buf, sizeof(cmd_buf), GFP_KERNEL);
+	if (!dmabuf) {
+		hid_err(hdev, "Failed to send type-cover hid report, couldn't allocate dmabuf\n");
+		return;
+	}
+
+	ret = hid_hw_raw_request(hdev, cmd_buf[0], dmabuf, sizeof(cmd_buf),
+				 HID_FEATURE_REPORT, HID_REQ_SET_REPORT);
+	kfree(dmabuf);
+
+	// FIXME: this should probably not log an error, there are races where it can happen but it's not bad
+	if (ret != sizeof (cmd_buf))
+		hid_err(hdev, "Failed to send type-cover hid report: %d\n", ret);
```

Comment:
Race with regards to what? The device state being suspended? That can be fixed with `pm_runtime_get`/`pm_runtime_put`.

Reply:
Yeah that was more a note to myself I guess, one thing I've seen is when closing the lid using the type cover which kinda cuts the usb connection (it's pretty weird, it's not disconnected, but the type cover turns off and doesn't send any more data, might be worth looking into, I think windows uses it to detect tablet mode somehow) and the request might not succeed.

Anyway, it doesn't seem to happen regularly, so I guess keeping the error message is fine.

Reply:
Hmm yeah, sounds a bit weird if the device itself is cutting the connection, but in that case we might not be able to do much about that.